tests/components/enphase_envoy/test_config_flow.py

```diff
@@ -397,6 +399,186 @@ async def test_zeroconf_host_already_exists(
     assert config_entry.title == "Envoy 1234"
 
 
+async def test_zero_conf_while_form(
+    hass: HomeAssistant, config_entry, setup_enphase_envoy
+) -> None:
+    """Test zeroconf while form is active."""
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
+    )
+    assert result["type"] is FlowResultType.FORM
+
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN,
+        context={"source": config_entries.SOURCE_ZEROCONF},
+        data=zeroconf.ZeroconfServiceInfo(
+            ip_address=ip_address("1.1.1.1"),
+            ip_addresses=[ip_address("1.1.1.1")],
+            hostname="mock_hostname",
+            name="mock_name",
+            port=None,
+            properties={"serialnum": "1234", "protovers": "7.0.1"},
+            type="mock_type",
+        ),
+    )
+    await hass.async_block_till_done()
+    assert result["type"] is FlowResultType.ABORT
+    assert config_entry.data["host"] == "1.1.1.1"
+    assert config_entry.unique_id == "1234"
+    assert config_entry.title == "Envoy 1234"
+
+
+async def test_zero_conf_second_envoy_while_form(
+    hass: HomeAssistant, config_entry, setup_enphase_envoy
+) -> None:
+    """Test zeroconf while form is active."""
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
+    )
+    assert result["type"] is FlowResultType.FORM
+
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN,
+        context={"source": config_entries.SOURCE_ZEROCONF},
+        data=zeroconf.ZeroconfServiceInfo(
+            ip_address=ip_address("4.4.4.4"),
+            ip_addresses=[ip_address("4.4.4.4")],
+            hostname="mock_hostname",
+            name="mock_name",
+            port=None,
+            properties={"serialnum": "4321", "protovers": "7.0.1"},
+            type="mock_type",
+        ),
+    )
+    await hass.async_block_till_done()
+    assert result["type"] is FlowResultType.FORM
+    assert config_entry.data["host"] == "1.1.1.1"
+    assert config_entry.unique_id == "1234"
+    assert config_entry.title == "Envoy 1234"
+
+
+async def test_zero_conf_malformed_serial_property(
+    hass: HomeAssistant, config_entry, setup_enphase_envoy
+) -> None:
+    """Test malformed zeroconf properties."""
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
+    )
+    assert result["type"] is FlowResultType.FORM
+
+    with pytest.raises(Exception):
```

Comment:
This looks like a B017 problem:

> B017: assertRaises(Exception) and pytest.raises(Exception) should be considered evil. They can lead to your test passing even if the code being tested is never executed due to a typo. Assert for a more specific exception (builtin or custom), or use assertRaisesRegex (if using assertRaises), or add the match keyword argument (if using pytest.raises), or use the context manager form with a target (e.g. with self.assertRaises(Exception) as ex:).

Reply:
added KeyError and assert for serialnumber in exception value